Block signals handlers when registering a new signal handler

iulianbarbu opened the pull request:
Fixes #52 

Blocked all signals handlers execution when registering
a signal handler, for the registered signal handler.
Refactored SignalHandler enum to a SignalHandler type.

Signed-off-by: Iulian Barbu <iul@amazon.com>

acatangiu commented:
I'm not sure there is a usecase to block specific signals while in a signal handler. I would just block all and explicitly declare that behaviour in the function documentation.

Re-entrant or cascaded signal handlers are usually discouraged.

iulianbarbu commented:
> I'm not sure there is a usecase to block specific signals while in a signal handler. I would just block all and explicitly declare that behaviour in the function documentation.
> 
> Re-entrant or cascaded signal handlers are usually discouraged.

I am fine with that. I will wait for more feedback though, then make the changes.

iulianbarbu commented:
@acatangiu , I've addressed your concerns. Please take a look: https://github.com/rust-vmm/vmm-sys-util/commit/8cd664d096cb2e346f03c0999e9f8e9a2ecfd2bf.

acatangiu commented on src/signal.rs:
```diff
@@ -183,21 +159,30 @@ pub fn validate_signal_num(num: c_int, for_vcpu: bool) -> errno::Result<c_int> {
+    // Safe, because this is a POD struct.
+    let mut act: sigaction = unsafe { mem::zeroed() };
+    act.sa_sigaction = handler as *const () as usize;
+    act.sa_flags = flag;
```
When using `sa_sigaction` instead of `sa_handler` to specify the handler, `sa_flags` needs to have `libc::SA_SIGINFO` set.

So at a minimum this line should be:
```rust
act.sa_flags = flag | libc::SA_SIGINFO
```

Looking at the other possible values of `sa_flags` I suggest we simply don't expose this as none of them seem like something VM's should use.

My preference is removing `flag` altogether from this function's param list and just set `sa_flags = SA_SIGINFO`.

acatangiu commented on src/signal.rs:
```diff
@@ -183,21 +159,30 @@ pub fn validate_signal_num(num: c_int, for_vcpu: bool) -> errno::Result<c_int> {
+        return errno::errno_result();
+    }
+
+    match unsafe { sigaction(num, &act, null_mut()) } {
```
nit: add a comment explaining why this is safe.

acatangiu commented on src/signal.rs:
```diff
@@ -183,21 +159,30 @@ pub fn validate_signal_num(num: c_int, for_vcpu: bool) -> errno::Result<c_int> {
+pub fn register_signal_handler(
     num: c_int,
     handler: SignalHandler,
     for_vcpu: bool,
```
I really think this `for_vcpu` is simply wrong. Like I said before, let's just remove it. You can do it in this PR or I can open a new one after, whichever you like.

acatangiu commented:
nit: suggest change in commit summary from:
```
- signal: block all sig hndlrs at sig handler regist
+ signal: block other signals during signal handling
```

acatangiu commented:
Also add entries to the `CHANGELOG` pertaining to the API changes.

iulianbarbu commented:
@acatangiu , I think I've included the feedback from the last review. Please take a look: https://github.com/rust-vmm/vmm-sys-util/commit/ab291fc935db89b94dd8d792952bfce7d14a81e3.

acatangiu commented on CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+`signal::validate_signal_num`, which puts inside the same bucket what
+we called `VCPU signals` and the others.
+* Removed `flag` arg of `signal::register_signal_handler` public methods,
+which defaults to `libc::SA_SIGINFO`.
```
```suggestion
which now defaults to `libc::SA_SIGINFO`.
```

acatangiu commented on CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+
+* Removed `for_vcpu` args from `signal::register_signal_handler` & from
+`signal::validate_signal_num`, which puts inside the same bucket what
+we called `VCPU signals` and the others.
```
I'd rephrase this to:
```
Removed `for_vcpu` argument from `signal::register_signal_handler` and
`signal::validate_signal_num`. Users now have full control which signals can
be used for their threads.
```

acatangiu commented on src/signal.rs:
```diff
@@ -129,32 +99,27 @@ fn SIGRTMAX() -> c_int {
-/// limits for realtime signals. The remaining ones need
+/// Usually, VCPU signals are assigned and must be enclosed within the OS
+/// limits for realtime signals, while the remaining ones need
 /// to be between the minimum (`SIGHUP`) and maximum (`SIGSYS`) values.
```
**must be enclosed** is incorrect.

Let's make the whole comment clearer:
```rust
/// Supported signals range from `SIGHUP` to `SIGUSR2` and from `SIGRTMIN` to `SIGRTMAX`.
/// We recommend using realtime signals `[SIGRTMIN, SIGRTMAX]` for VCPU threads.
```

iulianbarbu replied:
👍 done that.

acatangiu commented on src/signal.rs:
```diff
@@ -129,32 +99,27 @@ fn SIGRTMAX() -> c_int {
-        }
-    } else if SIGHUP <= num && num <= SIGSYS {
+pub fn validate_signal_num(num: c_int) -> errno::Result<c_int> {
+    if SIGHUP <= num && num <= SIGRTMAX() {
```
The `[SIGRTMIN, SIGRTMAX]` range varies depending on arch and libc implementation. There might be a gap between the 32 standard signal numbers and the 32 real-time signal numbers.

```suggestion
    if (SIGHUP <= num && num <= SIGUSR2) || (SIGRTMIN() <= num && num <= SIGRTMAX()) {
```

acatangiu commented on src/signal.rs:
```diff
@@ -538,22 +489,20 @@ mod tests {
         let killable = thread::spawn(|| loop {});
 
-        let res = killable.kill(SIGRTMAX());
+        let res = killable.kill(SIGRTMAX() + 1);
```
Is `SIGRTMAX()` usable? As in the first unusable is `SIGRTMAX()+1`?

iulianbarbu replied:
Hadn't found any resource that discourages usage of SIGRTMIN, respectively of SIGRTMAX, as signums for registration of signal handlers. The [signal(7)](http://man7.org/linux/man-pages/man7/signal.7.html) doc does not discourage the usage of SIGRTMIN, or SIGRTMAX, but expresses the need to use checks to avoid the case when `SIGRTMIN+n` exceeds `SIGRTMAX`.

As a seprate note, at this specific line, we have to use a signum which exceeds `SIGRTMAX()` because we look for the error thrown from `validate_signal_num` function.

This line could've been in the past as simple as:
 ```suggestion
        let res = killable.kill(16);
```

`SIGRTMAX()` usage, as an argument, is superfluous and does not mean that `killable.kill()` call will explicitly raise the `SIGRTMAX()` signal for the specific thread, but instead it would've raised the signal for the signum `SIGRTMIN() + SIGRTMAX()`, which is not a valid signal number also. While in the new case, with the simplifications from this PR, it is necessary to make the call for `killable.kill()` with an argument which explicitly exceeds `SIGRTMAX()`, so that is why I chose `SIGRTMAX() + 1`.

acatangiu commented on src/signal.rs:
```diff
@@ -129,32 +105,26 @@ fn SIGRTMAX() -> c_int {
-/// VCPU signals must be enclosed within the OS
-/// limits for realtime signals. The remaining ones need
-/// to be between the minimum (`SIGHUP`) and maximum (`SIGSYS`) values.
+/// Supported signals range from `SIGHUP` to `SIGUSR2` and from `SIGRTMIN` to `SIGRTMAX`.
```
```suggestion
/// Supported signals range from `SIGHUP` to `SIGSYS` and from `SIGRTMIN` to `SIGRTMAX`.
```

acatangiu commented on src/signal.rs:
```diff
@@ -129,32 +105,26 @@ fn SIGRTMAX() -> c_int {
-            return Ok(actual_num);
-        }
-    } else if SIGHUP <= num && num <= SIGSYS {
+pub fn validate_signal_num(num: c_int) -> errno::Result<c_int> {
```
Sorry for the late feedback, but since this function is now just validating and not mutating the signal number, how about we change it to returning a `bool` or a `Result<()>` (I prefer the latter):

```suggestion
pub fn validate_signal_num(num: c_int) -> errno::Result<()> {
```

iulianbarbu replied:
Yep, sure. I chose to return `Result<()>` because it is compatible with unit tests, and could be promptly returned in case of `Error`, while returning `bool` needs writing checks and explicit return of Error if the function call returns `false`.

acatangiu commented:
@sameo @jiangliu @andreeaflorescu please review.

aghecenco commented on src/signal.rs:
```diff
@@ -129,32 +105,26 @@ fn SIGRTMAX() -> c_int {
 ///
 /// # Arguments
 ///
 /// * `num`: the signal number to be verify.
```
nit
```suggestion
 /// * `num`: the signal number to be verified.
```

aghecenco commented on src/signal.rs:
```diff
@@ -183,21 +150,26 @@ pub fn validate_signal_num(num: c_int, for_vcpu: bool) -> errno::Result<c_int> {
+    act.sa_flags = libc::SA_SIGINFO;
+
+    // Block all signals while the `handler` is running.
+    // This makes async signal operations safe, because
```
nit: maybe explain that blocking other signals is needed to make sure the execution of the handler continues uninterrupted if another signal comes

aghecenco commented on CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+* Removed `for_vcpu` argument from `signal::register_signal_handler` and
+`signal::validate_signal_num`. Users now have full control which signals can
+be used for their threads.
+* Removed `flag` arg of `signal::register_signal_handler` public methods,
```
nit: you use `argument` above and `arg` here, please make them consistent

aghecenco commented on CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+
+* Removed `for_vcpu` argument from `signal::register_signal_handler` and
+`signal::validate_signal_num`. Users now have full control which signals can
+be used for their threads.
```
The users had full control before this change too, the syntax was just different:
```suggestion
`signal::validate_signal_num`. Users can now pass absolute values for all valid signal numbers.
```

acatangiu replied:
They didn't actually. At least not without "cheating". In the previous version, setting "for_vcpu" to `true` would restrict signal numbers to [SIGRTMIN, SIGRTMAX]. You could argue that one could use the function for setting up a signal handler for a vcpu but set `for_vcpu` to `false` but that's "cheating" 😛 

Either way, I'm not attached to this wording, I'm fine to change it, just wanted to give more context.

iulianbarbu commented:
@aghecenco , I've addressed all of your comments. Please take a look again at: https://github.com/rust-vmm/vmm-sys-util/commit/96ec93830ebf0c4ea3fb30f29ff4f3b93efed4bc.

acatangiu commented:
@rust-vmm/gatekeepers please review
